spring/integration/pubsub/src/main/java/com/example/spring/pubsub/PubsubApplication.java

```diff
@@ -0,0 +1,153 @@
+  }
+
+  @GetMapping("/listTopics")
+  public List<String> listTopics() {
+    return admin.listTopics().stream()
+        .map(Topic::getNameAsTopicName)
+        .map(TopicName::getTopic)
+        .collect(Collectors.toList());
+  }
+
+  @GetMapping("/listSubscriptions")
+  public List<String> listSubscriptions() {
+    return admin.listSubscriptions().stream()
+        .map(Subscription::getNameAsSubscriptionName)
+        .map(SubscriptionName::getSubscription)
+        .collect(Collectors.toList());
+  }
+
+  @PostMapping("/postMessage")
+  public RedirectView addMessage(@RequestParam("message") String message) {
+    messagingGateway.sendToPubsub(message);
+    return new RedirectView("/");
+  }
+
+  @PostMapping("/newTopic")
+  public RedirectView newTopic(@RequestParam("name") String topicName) {
+    admin.createTopic(topicName);
+    return new RedirectView("/");
+  }
+
+  @PostMapping("/newSubscription")
+  public RedirectView newSubscription(@RequestParam("name") String subscriptionName,
+      @RequestParam("topic") String topicName) {
+    admin.createSubscription(subscriptionName, topicName);
+    return new RedirectView("/");
+  }
+
+  @Bean
+  public MessageChannel pubsubInputChannel() {
+    return new PublishSubscribeChannel();
+  }
+
+  @Bean
+  public PubsubInboundChannelAdapter messageChannelAdapter(
+      @Qualifier("pubsubInputChannel") MessageChannel inputChannel,
+      SubscriberFactory subscriberFactory) {
+    PubsubInboundChannelAdapter adapter =
+        new PubsubInboundChannelAdapter(subscriberFactory, "messages");
+    adapter.setOutputChannel(inputChannel);
+    adapter.setAckMode(AckMode.MANUAL);
+
+    return adapter;
+  }
+
+  @Bean
+  @ServiceActivator(inputChannel = "pubsubInputChannel")
+  public MessageHandler receiveMessage() {
+    return message -> {
+      LOGGER.info("Message arrived! Payload: "
+          + ((ByteString) message.getPayload()).toStringUtf8());
+      AckReplyConsumer consumer = (AckReplyConsumer) message.getHeaders().get(
+          GcpHeaders.ACKNOWLEDGEMENT);
+      consumer.ack();
+    };
+  }
+
+  @Bean
+  @ServiceActivator(inputChannel = "pubsubInputChannel")
+  public MessageHandler receiveMessageInParallel() {
+    return message -> {
+      LOGGER.info("Message also arrived here! Payload: "
+          + ((ByteString) message.getPayload()).toStringUtf8());
+      AckReplyConsumer consumer = (AckReplyConsumer) message.getHeaders().get(
+          GcpHeaders.ACKNOWLEDGEMENT);
+      consumer.ack();
+    };
+  }
+
+  @Bean
+  @ServiceActivator(inputChannel = "pubsubOutputChannel")
+  public MessageHandler messageSender(PubsubTemplate pubsubTemplate) throws IOException {
+    PubsubMessageHandler outboundAdapter = new PubsubMessageHandler(pubsubTemplate);
+    outboundAdapter.setTopic("test");
+    return outboundAdapter;
+  }
+
+  @Bean
+  public MessageChannel pubsubOutputChannel() {
+    return new PublishSubscribeChannel();
```

Comment:
Really? I don't see more subscriber to this `pubsubOutputChannel`.
That is my concern to keep sample as simple as possible.

I agree about renaming subscribers to the `pubsubInputChannel` to avoid similar questions about parallelism